arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp

```diff
@@ -65,316 +65,270 @@
-      *coll, AccessMode::Type::WRITE);
-
-  Result res = trx.begin();
-
-  if (res.fail()) {
-    THROW_ARANGO_EXCEPTION(res);
-  }
-
-  for (auto const& link : links) {
-    if (link.second) {
-      // link excluded from recovery
-      _skippedIndexes.emplace(link.first->id());
-    } else {
-      // link participates in recovery
-      if (link.first->type() == Index::TRI_IDX_TYPE_INVERTED_INDEX) {
-        auto& impl =
-            basics::downCast<IResearchRocksDBInvertedIndex>(*(link.first));
-        impl.removeInRecovery(trx, docId, tick);
-      } else {
-        TRI_ASSERT(link.first->type() == Index::TRI_IDX_TYPE_IRESEARCH_LINK);
-        auto& impl = basics::downCast<IResearchRocksDBLink>(*(link.first));
-        impl.removeInRecovery(trx, docId, tick);
-      }
-    }
+  if (!links_needed &&
+      ranges.links.end >= static_cast<uint16_t>(_links.size())) {
+    _links.resize(ranges.links.begin);
+    _ranges[objectId].links = {};
   }
+}
 
-  res = trx.commit();
+void IResearchRocksDBRecoveryHelper::PutCF(uint32_t column_family_id,
+                                           rocksdb::Slice const& key,
+                                           rocksdb::Slice const& value,
+                                           rocksdb::SequenceNumber tick) {
+  return applyCF<true>(column_family_id, key, tick,
+                       [&](auto& impl, LocalDocumentId documentId) {
+                         auto doc = RocksDBValue::data(value);
+                         impl.recoveryInsert(tick, documentId, doc);
+                       });
+}
 
-  if (res.fail()) {
-    THROW_ARANGO_EXCEPTION(res);
-  }
+void IResearchRocksDBRecoveryHelper::DeleteCF(uint32_t column_family_id,
+                                              rocksdb::Slice const& key,
+                                              rocksdb::SequenceNumber tick) {
+  return applyCF<false>(column_family_id, key, tick,
+                        [&](auto& impl, LocalDocumentId documentId) {
+                          impl.recoveryRemove(documentId);
+                        });
 }
 
-void IResearchRocksDBRecoveryHelper::LogData(const rocksdb::Slice& blob,
+void IResearchRocksDBRecoveryHelper::LogData(rocksdb::Slice const& blob,
                                              rocksdb::SequenceNumber tick) {
-  RocksDBLogType const type = RocksDBLogValue::type(blob);
+  auto const type = RocksDBLogValue::type(blob);
 
   switch (type) {
-    case RocksDBLogType::IndexCreate: {
-      // Intentional NOOP. Index is committed upon creation.
-      // So if this marker was written  - index was persisted already.
+    case RocksDBLogType::DatabaseDrop:
+    case RocksDBLogType::CollectionDrop:
+      // TODO(MBkkt) Do we really need clear for these?
+    case RocksDBLogType::DatabaseCreate:
+    case RocksDBLogType::CollectionCreate:
+    case RocksDBLogType::IndexCreate:
+    case RocksDBLogType::IndexDrop:
+      // TODO(MBkkt) Clear only necessary entries
```

Comment:
I mean we could get objectId and remove it from _ranges.
```cpp
auto const objectId = RocksDBLogValue::objectId(blob);
_ranges.erase(objectId);
```
But I don't sure will something like this correct, or it could return something invalid

Reply:
Looks like only `case RocksDBLogType::IndexCreate:` makes sense

Reply:
Done